bk72xx - call of overloaded 'millis()' is ambiguous

### The problem

This is a re-report of esphome/issues#6118 (from the old issues repo).
The problem is triggered when trying to use millis() in a lambda but only on bk72xx platforms, because it is in conflict with the plain-C function with the same name that resides somewhere in the arduino framework.
```yaml
bk72xx:

wifi:
  on_connect:
    then:
      - lambda: |-
          id(connect_count) += 1.0f;
          id(connect_count_sensor).publish_state(id(connect_count));

          float connection_time_s = (millis() - id(disconnect_ts)) / 1000.0f;
          id(connection_time).publish_state(connection_time_s);

  on_disconnect:
    then:
      - lambda: |-
          id(disconnect_ts) = millis();
```

```
/config/remote/../common/wifi_no_ip.yaml:34:41: error: call of overloaded 'millis()' is ambiguous
   34 |           float connection_time_s = (millis() - id(disconnect_ts)) / 1000.0f;
      |                                         ^
In file included from /config/.esphome/platformio/packages/framework-arduino-api/api/Interrupts.h:8,
                 from /config/.esphome/platformio/packages/framework-arduino-api/api/ArduinoAPI.h:29,
                 from /libretiny/cores/common/arduino/src/Arduino.h:25,
                 from <command-line>:
/config/.esphome/platformio/packages/framework-arduino-api/api/Common.h:103:15: note: candidate: 'long unsigned int millis()'
  103 | unsigned long millis(void);
      |               ^~~~~~
In file included from src/esphome/core/application.h:9,
                 from src/esphome/components/api/api_frame_helper.h:13,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/hal.h:42:10: note: candidate: 'uint32_t esphome::millis()'
   42 | uint32_t millis();
      |          ^~~~~~
```



Using **esphome::millis()** instead solves it, but this in turn triggers compiler error on other platforms.
Here is the error **for esp8266**:

```
/config/heater/../common/wifi_no_ip.yaml: In lambda function:
src/main.cpp:75:18: error: 'esphome::esphome' has not been declared
   75 | #define millis() esphome::millis()
      |                  ^~~~~~~
/config/heater/../common/wifi_no_ip.yaml:34:43: note: in expansion of macro 'millis'
   34 |           float connection_time_s = (esphome::millis() - id(disconnect_ts)) / 1000.0f;
      |                                           ^~~~~~
```

It turns out this is caused by some very old workaround, in config.py, that **originally was supposed to fix ambiguities like that**. Maybe it's not working as it should in case of bk72xx?

```
ARDUINO_GLUE_CODE = """\
#define yield() esphome::yield()
#define millis() esphome::millis()
#define micros() esphome::micros()
#define delay(x) esphome::delay(x)
#define delayMicroseconds(x) esphome::delayMicroseconds(x)
"""

```


I use the lambda part in a package that is common for all my devices and due to this problem I will probably need to split it into separate packages for bk72xx and the rest. I would like to avoid that....

### Which version of ESPHome has the issue?

2025.12.0-dev

### What type of installation are you using?

Docker

### What platform are you using?

BK72XX

### Component causing the issue

core

### YAML Config

```yaml

```

### Anything in the logs that might be useful for us?

```txt

```

### Additional information

_No response_

--- a/esphome/core/config.py
+++ b/esphome/core/config.py
@@ -382,10 +382,15 @@ def include_file(path: Path, basename: Path, is_c_header: bool = False):
 
 
 ARDUINO_GLUE_CODE = """\
+#undef yield
 #define yield() esphome::yield()
+#undef millis
 #define millis() esphome::millis()
+#undef micros
 #define micros() esphome::micros()
+#undef delay
 #define delay(x) esphome::delay(x)
+#undef delayMicroseconds
 #define delayMicroseconds(x) esphome::delayMicroseconds(x)
 """
 

--- a/esphome/core/config.py
+++ b/esphome/core/config.py
@@ -537,7 +537,7 @@ async def to_code(config: ConfigType) -> None:
     if config[CONF_DEBUG_SCHEDULER]:
         cg.add_define("ESPHOME_DEBUG_SCHEDULER")
 
-    if CORE.using_arduino and not CORE.is_bk72xx:
+    if CORE.using_arduino:
         CORE.add_job(add_arduino_global_workaround)
 
     if config[CONF_INCLUDES]:

